app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt

```diff
@@ -0,0 +1,134 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.raulh82vlc.functionalkotlinarrow.data.network
+
+import arrow.Kind
+import arrow.core.Try
+import arrow.core.right
+import arrow.effects.IO
+import arrow.effects.async
+import arrow.effects.fix
+import arrow.effects.monadError
+import arrow.effects.typeclasses.Async
+import arrow.typeclasses.binding
+import com.google.gson.Gson
+import com.jakewharton.retrofit2.adapter.kotlin.coroutines.experimental.CoroutineCallAdapterFactory
+import com.raulh82vlc.functionalkotlinarrow.data.NetworkToCacheMapper
+import com.raulh82vlc.functionalkotlinarrow.data.cache.model.FeedItemCacheModel
+import com.raulh82vlc.functionalkotlinarrow.data.network.error.ApiException
+import com.raulh82vlc.functionalkotlinarrow.data.network.model.FeedApiModel
+import kotlinx.coroutines.experimental.CommonPool
+import okhttp3.OkHttpClient
+import okhttp3.ResponseBody
+import retrofit2.Call
+import retrofit2.Response
+import retrofit2.Retrofit
+import java.io.IOException
+import java.util.concurrent.TimeUnit
+
+object NetworkDataSource {
+
+    private const val JSON_FLICKR_FEED_LITERAL = "jsonFlickrFeed("
+    private const val JSON_FLICKR_FEED_KEYWORD = "jsonFlickrFeed\\("
+    private const val API_HOST = "https://api.flickr.com/services/"
+
+    val gson = Gson()
+
+    val retrofit = Retrofit.Builder()
+            .baseUrl(API_HOST)
+            .client(createOkHttp())
+            .addCallAdapterFactory(CoroutineCallAdapterFactory())
+            .build()
+            .create(FeedApi::class.java)
+
+    private fun createOkHttp(): OkHttpClient {
+        val okHttpClient = OkHttpClient.Builder()
+                .connectTimeout(3, TimeUnit.SECONDS)
+                .writeTimeout(10, TimeUnit.SECONDS)
+                .readTimeout(10, TimeUnit.SECONDS)
+        return okHttpClient.build()
+    }
+
+    fun requestFeed(format: String): IO<List<FeedItemCacheModel>> {
+        val monadError = IO.monadError()
+        return monadError.binding {
+            val result = runInAsyncContext(
+                    f = { transformToRepository(retrofit.getFeed(format)) },
+                    onError = { monadError.raiseError<List<FeedItemCacheModel>>(it) },
+                    onSuccess = { monadError.just(it) },
+                    AC = IO.async()
+            ).bind()
+            result.bind()
+        }.fix()
+    }
+
+    private fun transformToRepository(feed: Call<ResponseBody>): List<FeedItemCacheModel> {
+        val response = execute(feed)?.string()
+        response?.run {
+            if (hasFeedFormat(this)) {
+                return NetworkToCacheMapper.map(deserializeFeedJSON(extractJSONFromResponse(this)).feedItems)
+            }
+            return emptyList()
+        }
+        return emptyList()
+    }
+
+    private fun hasFeedFormat(responseBodyResult: String): Boolean =
```

Comment:
This can have a `String` as the receiver so can do `string.hasFeedFormat()`

Reply:
like an extension function you mean?